In [1]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, ConfusionMatrixDisplay
from sklearn import tree
from matplotlib import pyplot as plt
import pandas as pd

In [2]:
load_diabetes().DESCR

'.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n:Number of Instances: 442\n\n:Number of Attributes: First 10 columns are numeric predictive values\n\n:Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n:Attribute Information:\n    - age     age in years\n    - sex\n    - bmi     body mass index\n    - bp      average blood pressure\n    - s1      tc, total serum cholesterol\n    - s2      ldl, low-density lipoproteins\n    - s3      hdl, high-density lipoproteins\n    - s4      tch, total cholesterol / HDL\n    - s5      ltg, possibly log of serum triglycerides level\n    - s6      glu, blood sugar level\n\nNote: Each of thes

'.. _diabetes_dataset: набор данных о диабете
 для каждого из n =442 пациентов с сахарным диабетом были получены все базовые переменные, возраст, пол, индекс массы тела, среднее артериальное давление и шесть показателей сыворотки крови пациентов, а также интересующая реакция являются количественным показателем прогрессирования заболевания через год после исходного уровня.
 ** Характеристики набора данных:**
 Количество экземпляров: 442
 Количество атрибутов: Первые 10 столбцов - числовые прогностические значения\n\n: Цель: Столбец 11 - количественный показатель прогрессирования заболевания через год после исходного уровня 
 Информация об атрибуте:
     - возраст возраст в годах
    - пол 
    - имт индекс массы тела
    - ад среднее артериальное давление
    - s2 ЛПНП, липопротеины низкой плотности
    - s3 лпвп, липопротеины высокой плотности
    - s4 tc, общий холестерин / Уровень ЛПВП 
    - s5 ltg, возможно, логарифм уровня триглицеридов в сыворотке крови 
    - s6 glu, уровень сахара в крови 
    Примечание: Каждая из этих 10 характерных переменных была центрирована по среднему значению и масштабирована на стандартное отклонение, умноженное на квадратный корень из `n_samples` (т.е. сумма квадратов каждого столбца равна 1).URL-адрес источника:https://www4.stat.ncsu.edu/~boos/var.select/diabetes.html Для получения дополнительной информации смотрите: Брэдли Эфрон, Тревор Хасти, Иэн Джонстон и Роберт Тибширани. (2004) "Регрессия под наименьшим углом", Annals of Statistics (с обсуждением), 407-499.\n (https://web.stanford.edu/~hastie/Papers/LARS/LeastAngle_2002.pdf)\n'

In [3]:
df=load_diabetes(as_frame=True).frame


In [4]:
df.describe()

age           sex           bmi            bp            s1  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean  -2.511817e-19  1.230790e-17 -2.245564e-16 -4.797570e-17 -1.381499e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.072256e-01 -4.464164e-02 -9.027530e-02 -1.123988e-01 -1.267807e-01   
25%   -3.729927e-02 -4.464164e-02 -3.422907e-02 -3.665608e-02 -3.424784e-02   
50%    5.383060e-03 -4.464164e-02 -7.283766e-03 -5.670422e-03 -4.320866e-03   
75%    3.807591e-02  5.068012e-02  3.124802e-02  3.564379e-02  2.835801e-02   
max    1.107267e-01  5.068012e-02  1.705552e-01  1.320436e-01  1.539137e-01   

                 s2            s3            s4            s5            s6  \
count  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02  4.420000e+02   
mean   3.918434e-17 -5.777179e-18 -9.042540e-18  9.293722e-17  1.130318e-17   
std    4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02  4.761905e-02   
min   -1.156131e-01 -1.023071e-01 -7.639450e-02 -1.260971e-01 -1.377672e-01   
25%   -3.035840e-02 -3.511716e-02 -3.949338e-02 -3.324559e-02 -3.317903e-02   
50%   -3.819065e-03 -6.584468e-03 -2.592262e-03 -1.947171e-03 -1.077698e-03   
75%    2.984439e-02  2.931150e-02  3.430886e-02  3.243232e-02  2.791705e-02   
max    1.987880e-01  1.811791e-01  1.852344e-01  1.335973e-01  1.356118e-01   

           target  
count  442.000000  
mean   152.133484  
std     77.093005  
min     25.000000  
25%     87.000000  
50%    140.500000  
75%    211.500000  
max    346.000000

In [5]:
df['target'].value_counts()

target
200.0    6
72.0     6
178.0    5
71.0     5
90.0     5
        ..
146.0    1
212.0    1
120.0    1
183.0    1
57.0     1
Name: count, Length: 214, dtype: int64

In [6]:
features=df.drop(['target'],axis=1)
target=df['target']
print(features.shape)
print(target.shape)

(442, 10)
(442,)


In [7]:
features_train, features_val, target_train, target_val=train_test_split(features, target, test_size=0.3, random_state=45)

In [8]:
target_train.value_counts(normalize=True).sort_index()

target
31.0     0.003236
37.0     0.003236
39.0     0.003236
40.0     0.003236
42.0     0.009709
           ...   
317.0    0.003236
321.0    0.003236
332.0    0.003236
336.0    0.003236
341.0    0.003236
Name: proportion, Length: 181, dtype: float64

In [9]:
target_val.value_counts(normalize=True).sort_index()

target
25.0     0.007519
39.0     0.007519
45.0     0.007519
53.0     0.007519
59.0     0.007519
           ...   
293.0    0.007519
302.0    0.007519
303.0    0.007519
306.0    0.007519
346.0    0.007519
Name: proportion, Length: 96, dtype: float64

In [10]:
model=tree.DecisionTreeClassifier()
model.fit(features_train, target_train)


DecisionTreeClassifier()

In [11]:
predictions_val=model.predict(features_val)
def count_errors(true_answers, pred_answers):
    all_answers_together=zip(true_answers, pred_answers)
    errors_list=[1 if v[0]!=v[1] else 0 for v in all_answers_together]
    return sum(errors_list)
print('Ошибок:', count_errors(target_val, predictions_val))

Ошибок: 132


In [14]:
from sklearn.linear_model import LinearRegression


In [15]:
model2=LinearRegression()
model2.fit(features_train, target_train)

LinearRegression()

In [17]:
predictions_val2=model2.predict(features_val)
def count_errors(true_answers, pred_answers):
    all_answers_together=zip(true_answers, pred_answers)
    errors_list=[1 if v[0]!=v[1] else 0 for v in all_answers_together]
    return sum(errors_list)
print('Ошибок:', count_errors(target_val, predictions_val))

Ошибок: 132
